versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/KeyListBuildState.java

```diff
@@ -67,62 +60,50 @@ class KeyListBuildState {
     this.serializedEntrySize = serializedEntrySize;
   }
 
-  void add(KeyListEntry keyListEntry) {
-    int entrySize = serializedEntrySize.applyAsInt(keyListEntry);
-    if (embedded) {
-      // filling the embedded key-list in CommitLogEntry
-
-      if (currentSize + entrySize < maxEmbeddedKeyListSize) {
-        // CommitLogEntry.keyList still has room
-        addToEmbedded(keyListEntry, entrySize);
-      } else {
-        // CommitLogEntry.keyList is "full", switch to the first KeyListEntity
-        embedded = false;
-        newKeyListEntity();
-        addToKeyListEntity(keyListEntry, entrySize);
-      }
-    } else {
-      // filling linked key-lists via CommitLogEntry.keyListIds
-
-      if (currentSize + entrySize > maxKeyListEntitySize) {
-        // current KeyListEntity is "full", switch to a new one
-        finishKeyListEntity();
-        newKeyListEntity();
-      }
-      addToKeyListEntity(keyListEntry, entrySize);
-    }
+  void add(KeyListEntry entry) {
+    entries.add(entry);
   }
 
-  private void finishKeyListEntity() {
-    Hash id = randomHash();
-    newKeyListEntities.add(KeyListEntity.of(id, currentKeyList.build()));
-    newCommitEntry.addKeyListsIds(id);
+  int bucket(KeyListEntry entry, int bucketCount) {
+    return Math.abs(entry.getKey().hashCode() % bucketCount);
   }
 
-  private void newKeyListEntity() {
-    currentSize = 0;
-    currentKeyList = ImmutableKeyList.builder();
-  }
+  List<KeyListEntity> finish() {
+    int totalSize = entries.stream().mapToInt(serializedEntrySize).sum();
+    int remainingEmbedded = maxEmbeddedKeyListSize - currentSize;
+    int maxBucketSize =
+        Math.max(MINIMUM_BUCKET_SIZE, Math.min(remainingEmbedded, maxKeyListEntitySize));
 
-  private void addToKeyListEntity(KeyListEntry keyListEntry, int keyTypeSize) {
-    currentSize += keyTypeSize;
-    currentKeyList.addKeys(keyListEntry);
-  }
+    int bucketCount = totalSize / maxBucketSize + 1;
+    List<List<KeyListEntry>> buckets = new ArrayList<>(bucketCount);
+    for (int i = 0; i < bucketCount; i++) {
+      buckets.add(new ArrayList<>());
+    }
```

Comment:
I think buckets can still overflow in terms of "entity size" in this case if the key distribution is not strictly uniform. Would it make sense to start low and increase the bucket count iteratively until all buckets fit under the size limit?

Alternatively, this seems like a reasonable case for open hashing... That would be more flexible in terms of incremental inserts. WDYT?

Reply:
> buckets can still overflow in terms of "entity size"

they actually do - both due to length of the keys _and_ unlucky hash distribution (buckets w/ 0 entries .. buckets with 3x more entries).